File: 383_ransom_note/383_ransom_note.ts
function canConstruct(ransomNote: string, magazine: string): boolean {
  let constructedMap: Map<string, number> = new Map<string, number>();
  for (const char of magazine) {
    constructedMap.set(char, (constructedMap.get(char) || 0) + 1);
  }
  for (const char of ransomNote) {
    constructedMap.set(char, (constructedMap.get(char) || 0) - 1);
    if (constructedMap.get(char) || 0 < 0) {
      return false;
    }
  }
  return true;
}

File: 383_ransom_note/383_ransom_note.java
import java.util.HashMap;

class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        HashMap<Character, Integer> constructedMap = new HashMap<>();
        for (char chr : magazine.toCharArray()) {
            constructedMap.put(chr, constructedMap.getOrDefault(chr, 0) + 1);
        }
        for (char chr : ransomNote.toCharArray()) {
            constructedMap.put(chr, constructedMap.get(chr));
            if (constructedMap.getOrDefault(chr, 0) < 0) {
                return false;
            }
        }
        return true;
    }
}